src/opm/parser/eclipse/EclipseState/NumericalAquifer.cpp

```diff
@@ -0,0 +1,327 @@
+            }
+            aquifer.addAquiferConnection(con.second);
+            con_set.insert(con_global_index);
+        }
+    }
+}
+
+bool NumericalAquifers::empty() const {
+    return this->aquifers_.empty();
+}
+
+void NumericalAquifers::updateCellProps(const EclipseGrid& grid,
+                                        std::vector<double>& pore_volume,
+                                        std::vector<int>& satnum,
+                                        std::vector<int>& pvtnum,
+                                        std::vector<double>& cell_depth) const {
+    for (const auto& iter : this->aquifers_) {
+        iter.second.updateCellProps(grid, pore_volume, satnum, pvtnum, cell_depth);
+    }
+}
+
+std::array<std::set<size_t>, 3>
+NumericalAquifers::transToRemove(const EclipseGrid& grid) const {
+    std::array<std::set<size_t>, 3> trans;
+    for (const auto& pair : this->aquifers_) {
+        auto trans_aquifer = pair.second.transToRemove(grid);
+        for (size_t i = 0; i < 3; ++i) {
+            trans[i].merge(trans_aquifer[i]);
+        }
+    }
+    return  trans;
+}
+
+void NumericalAquifers::
+appendNNC(const EclipseGrid &grid, const FieldPropsManager &fp, NNC &nnc) const {
+    for (const auto& pair : this->aquifers_) {
+        pair.second.appendNNC(grid, fp, nnc);
+    }
+}
+
+const std::unordered_map<size_t, const NumericalAquiferCell>& NumericalAquifers::
+aquiferCells() const {
+    return this->aquifer_cells_;
+}
+
+bool NumericalAquifers::hasCell(const size_t cell_global_index) const {
+    const auto& cells = this->aquifer_cells_;
+    return (cells.find(cell_global_index) != cells.end());
+}
+
+const NumericalAquiferCell& NumericalAquifers::getCell(const size_t cell_global_index) const {
+    assert(this->hasCell(cell_global_index));
+    return this->aquifer_cells_.at(cell_global_index);
+}
+
+using AQUNUM = ParserKeywords::AQUNUM;
+NumericalAquiferCell::NumericalAquiferCell(const DeckRecord& record, const EclipseGrid& grid, const FieldPropsManager& field_props)
+   : aquifer_id( record.getItem<AQUNUM::AQUIFER_ID>().get<int>(0) )
+   , I ( record.getItem<AQUNUM::I>().get<int>(0) - 1 )
+   , J ( record.getItem<AQUNUM::J>().get<int>(0) - 1 )
+   , K ( record.getItem<AQUNUM::K>().get<int>(0) - 1 )
+   , area (record.getItem<AQUNUM::CROSS_SECTION>().getSIDouble(0) )
+   , length ( record.getItem<AQUNUM::LENGTH>().getSIDouble(0) )
+   , permeability( record.getItem<AQUNUM::PERM>().getSIDouble(0) )
+{
+    const auto& cell_depth = field_props.cellDepth();
+    const auto& poro = field_props.get_double("PORO");
+    const auto& pvtnum = field_props.get_int("PVTNUM");
+    const auto& satnum = field_props.get_int("SATNUM");
+
+    this->global_index = grid.getGlobalIndex(I, J, K);
+
+    if ( !record.getItem<AQUNUM::PORO>().defaultApplied(0) ) {
+        this->porosity = record.getItem<AQUNUM::PORO>().getSIDouble(0);
+    } else {
+        this->porosity = poro[global_index];
```

Comment:
The vectors from `field_props` only have active elements; i.e. you must use `active_index` when looking up values here.

Reply:
You are right.